fix(formatter): Improve AssignmentExpression parentheses handling

leaysgur opened the pull request:
For this input given:

```js
async function m() {
  ({ x } = await foo());
  ({ y } = foo());
}
```

Currently formatted as:

```js
async function m() {
  { x } = await foo();
  { y } = foo();
}
```

This is invalid syntax.

Dunqing reviewed (changes requested):
Great, I like the with detailed comments. CI fails, and I can see that one of the failed tests is related to `ArrowFunctionExpression`. I've explained the reason below.

Dunqing reviewed (approved):
Great!

leaysgur committed:
fix(formatter): Improve AssignmentExpression parentheses handling (#13668)

For this input given:

```js
async function m() {
  ({ x } = await foo());
  ({ y } = foo());
}
```

Currently formatted as:

```js
async function m() {
  { x } = await foo();
  { y } = foo();
}
```

This is invalid syntax.
